maps_dashboards/public/components/map_page/map_page.tsx

```diff
@@ -3,33 +3,51 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
-import React, { useEffect, useState } from 'react';
+import React, { useEffect, useRef, useState } from 'react';
 import { useParams } from 'react-router-dom';
 import { SimpleSavedObject } from 'opensearch-dashboards/public';
+import { Map as Maplibre } from 'maplibre-gl';
 import { MapContainer } from '../map_container';
 import { MapTopNavMenu } from '../map_top_nav';
 import { MapLayerSpecification } from '../../model/mapLayerType';
 import { MapServices } from '../../types';
 import { useOpenSearchDashboards } from '../../../../../src/plugins/opensearch_dashboards_react/public';
 import { MapSavedObjectAttributes } from '../../../common/map_saved_object_attributes';
-import { OPENSEARCH_MAP_LAYER } from '../../../common';
-import { getLayerConfigMap } from '../../utils/getIntialLayerConfig';
+import { DASHBOARDS_MAPS_LAYER_TYPE, OPENSEARCH_MAP_LAYER } from '../../../common';
+import { getLayerConfigMap, getInitialMapState } from '../../utils/getIntialConfig';
+import { IndexPattern } from '../../../../../src/plugins/data/public';
+import { MapState } from '../../model/mapState';
 
 export const MapPage = () => {
-  const [layers, setLayers] = useState<MapLayerSpecification[]>([]);
-  const { id: mapIdFromUrl } = useParams<{ id: string }>();
-  const [savedMapObject, setSavedMapObject] =
-    useState<SimpleSavedObject<MapSavedObjectAttributes> | null>();
   const { services } = useOpenSearchDashboards<MapServices>();
   const {
     savedObjects: { client: savedObjectsClient },
   } = services;
+  const [layers, setLayers] = useState<MapLayerSpecification[]>([]);
+  const { id: mapIdFromUrl } = useParams<{ id: string }>();
+  const [savedMapObject, setSavedMapObject] =
+    useState<SimpleSavedObject<MapSavedObjectAttributes> | null>();
+  const [layersIndexPatterns, setLayersIndexPatterns] = useState<IndexPattern[]>([]);
+  const maplibreRef = useRef<Maplibre | null>(null);
+  const [mapState, setMapState] = useState<MapState>(getInitialMapState());
 
   useEffect(() => {
     if (mapIdFromUrl) {
       savedObjectsClient.get<MapSavedObjectAttributes>('map', mapIdFromUrl).then((res) => {
         setSavedMapObject(res);
-        setLayers(JSON.parse(res.attributes.layerList as string));
+        const layerList: MapLayerSpecification[] = JSON.parse(res.attributes.layerList as string);
+        const savedMapState: MapState = JSON.parse(res.attributes.mapState as string);
```

Comment:
Type casting is potentially dangerous for runtime, I got a runtime error on this line because I have previously created map which `mapState` is `undefined`

Reply:
@ruanyl Thanks call out here! I think previously in your local code commit, there didn't have mapState. Currently for all initially created maps, it will have mapState property to the map saved object.